### DISCUSSION TITLE
In the SPA example, if the menu one button could change state when pressed

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#2961] [Q&A] [answered]
Can the menu button change its state, such as bgcolor, according to the current page that is clicked?

![微信截图_20240427083220](https://github.com/zauberzeug/nicegui/assets/25096045/4a8174f3-8da0-439e-9365-303008f0aec9)

**Comment by falkoschindler** [comment, marked as answer]
Hi @v479038280,

Here is a possible solution:

```py
from router import Router
from nicegui import ui

@ui.page('/')
@ui.page('/{path:path}')
def main(path: str = ''):
    router = Router()

    @router.add('/')
    def show_one():
        ui.label('Content One').classes('text-2xl')

    @router.add('/two')
    def show_two():
        ui.label('Content Two').classes('text-2xl')

    @router.add('/three')
    def show_three():
        ui.label('Content Three').classes('text-2xl')

    def open_route(target: str):
        router.open(target)
        navigation.refresh(target)

    @ui.refreshable
    def navigation(path: str):
        with ui.row():
            ui.button('One', on_click=lambda: open_route('/')).props('color=orange' if path == '/' else '')
            ui.button('Two', on_click=lambda: open_route('/two')).props('color=orange' if path == '/two' else '')
            ui.button('Three', on_click=lambda: open_route('/three')).props('color=orange' if path == '/three' else '')
    navigation(f'/{path}')

    router.frame().classes('w-full p-4 bg-gray-100')

ui.run()
```

<img width="487" alt="Screenshot 2024-04-27 at 12 07 54" src="https://github.com/zauberzeug/nicegui/assets/5767091/4018895c-a589-43f1-9ade-169c2fec8aee">

**Reply by v479038280** [reply]
This one could replace the one in the example

**Comment by v479038280** [comment]
> Hi @v479038280,
> 
> Here is a possible solution:
> 
> ```python
> from router import Router
> from nicegui import ui
> 
> @ui.page('/')
> @ui.page('/{path:path}')
> def main(path: str = ''):
>     router = Router()
> 
>     @router.add('/')
>     def show_one():
>         ui.label('Content One').classes('text-2xl')
> 
>     @router.add('/two')
>     def show_two():
>         ui.label('Content Two').classes('text-2xl')
> 
>     @router.add('/three')
>     def show_three():
>         ui.label('Content Three').classes('text-2xl')
> 
>     def open_route(target: str):
>         router.open(target)
>         navigation.refresh(target)
> 
>     @ui.refreshable
>     def navigation(path: str):
>         with ui.row():
>             ui.button('One', on_click=lambda: open_route('/')).props('color=orange' if path == '/' else '')
>             ui.button('Two', on_click=lambda: open_route('/two')).props('color=orange' if path == '/two' else '')
>             ui.button('Three', on_click=lambda: open_route('/three')).props('color=orange' if path == '/three' else '')
>     navigation(f'/{path}')
> 
>     router.frame().classes('w-full p-4 bg-gray-100')
> 
> ui.run()
> ```
> 
> <img alt="Screenshot 2024-04-27 at 12 07 54" width="487" src="https://private-user-images.githubusercontent.com/5767091/326170822-4018895c-a589-43f1-9ade-169c2fec8aee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyNzcxMTEsIm5iZiI6MTcxNDI3NjgxMSwicGF0aCI6Ii81NzY3MDkxLzMyNjE3MDgyMi00MDE4ODk1Yy1hNTg5LTQzZjEtOWFkZS0xNjljMmZlYzhhZWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjhUMDQwMDExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjI5ZmJiZTliNTViNTQxMWZjMTliMTcyNjIzMDcxZWM1M2I1YzY2MWRlYWIzZTQyZTBhNGFlMWQ5MzY3MGJkNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tdclzRUOjB2CLKe0kEbSqRrwkjEa4ST4aLwQZNLAHb8">

Is't possible to change the state of ui.link('a','pagea') correspond to the opend page?  if path==current url?

**Reply by v479038280** [reply]
const fullURL = window.location.href; 
url could get by this way

**Reply by falkoschindler** [reply]
Sorry, I don't know what you mean. What "state" do you want to change? Which `ui.link` are you referring to? Why do you want to use `window.location.href` instead of the `path` attribute?
I think my example gives a good starting point to change the appearance of the navigation element. Of course you can replace buttons with links and represent the current state of each item differently.

**Reply by v479038280** [reply]
May be I mean " Is there a way to implement a method similar to changing the button background color in an interface with a shared navigation bar created using the @contextmanager?"  translated by AI ,better than me 😃

I find a problem in the SPA, only the button in  left_drawer could change color ,right_drawer doesn't work, also the ui.header()
```py
@ui.refreshable
def navigation(path: str):
    with ui.header():
        ui.label('Navigation').classes('text-2xl')
        ui.button('One', on_click=lambda: open_route('/')).props('color=orange' if path == '/' else '')
        ui.button('Two', on_click=lambda: open_route('/two')).props('color=orange' if path == '/two' else '')
        ui.button('Three').props('color=orange' if path == '/three' else '')
    with ui.left_drawer().classes('items-start p-10').props('width=200'):
        ui.button('One', on_click=lambda: open_route('/')).props('color=orange' if path == '/' else '')
        ui.button('Two', on_click=lambda: open_route('/two')).props('color=orange' if path == '/two' else '')
        ui.button('Three', on_click=lambda: open_route('/three')).props('color=orange' if path == '/three' else '')
        a=ui.label('one').on('click', lambda: open_route('/')).classes('bg-orange-500' if path == '/' else '')
        a.classes('cursor-pointer')
        ui.label('two').on('click', lambda: open_route('/two')).classes('bg-orange-500' if path == '/two' else '')
        ui.label('three').on('click', lambda: open_route('/three')).classes('bg-orange-500' if path == '/three' else '')
    with ui.right_drawer().classes('items-start p-10').props('width=200'):
        ui.button('One', on_click=lambda: open_route('/')).props('color=orange' if path == '/' else '')
        ui.button('Two', on_click=lambda: open_route('/two')).props('color=orange' if path == '/two' else '')
        ui.button('Three', on_click=lambda: open_route('/three')).props('color=orange' if path == '/three' else '')
        a=ui.label('one').on('click', lambda: open_route('/')).classes('bg-orange-500' if path == '/' else '')
        a.classes('cursor-pointer')
        ui.label('two').on('click', lambda: open_route('/two')).classes('bg-orange-500' if path == '/two' else '')
        ui.label('three').on('click', lambda: open_route('/three')).classes('bg-orange-500' if path == '/three' else '')
```

**Reply by falkoschindler** [reply]
There is an error:
> Found top level layout element "Header" inside element "RefreshableContainer". Top level layout elements should not be nested but must be direct children of the page content. This will be raising an exception in NiceGUI 1.5

These top-level layout element are not supposed to be nested in other elements like a refreshable container. Nest the refreshable instead:
```py
@ui.refreshable
def navigation(path: str):
    ui.button('One', on_click=lambda: open_route('/')).props('color=orange' if path == '/' else '')
    ui.button('Two', on_click=lambda: open_route('/two')).props('color=orange' if path == '/two' else '')
    ui.button('Three', on_click=lambda: open_route('/three')).props('color=orange' if path == '/three' else '')
with ui.header():
    navigation(f'/{path}')
```
This works for drawer as well.
